### PR TITLE
Add precision inference

### PR DESCRIPTION
**Pull request by zhiqiangdon**
Add one util function to infer precision. Reduce boilerplate code.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: multimodal/src/autogluon/multimodal/utils.py [hunk 1]
```diff
@@ -2693,3 +2693,40 @@ def from_coco(
 
 def getCOCOCatIDs():
     return [e for e in range(1, 91) if e not in {12, 26, 29, 30, 45, 66, 68, 69, 71, 83}]
+
+
+def infer_precision(num_gpus: int, precision: Union[int, str]):
+    """
+    Infer the precision based on the provided and the environment.
```
Comment (suzhoum) [inline multimodal/src/autogluon/multimodal/utils.py:2700]: Did you mean "based on the provided environment?"

**Review by suzhoum — APPROVED** [review]
Thanks for the refactor! Just left a few minor comments on the wordings.

File: multimodal/src/autogluon/multimodal/utils.py (hunk 1 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/utils.py]: Updated. I have made it more clear with "the provided precision".

File: multimodal/src/autogluon/multimodal/utils.py [hunk 2]
```diff
@@ -2693,3 +2693,40 @@ def from_coco(
...
+    Returns
+    -------
+    The inferred precision.
+    """
+    if num_gpus == 0:  # CPU only prediction
+        warnings.warn(
+            "Only CPU is detected in the instance. "
+            "MultiModalPredictor will predict with CPU only. "
+            "This may results in slow speed. "
+            "Consider to switch to an instance with GPU support.",
```
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/utils.py]: Thanks! Corrected!

File: multimodal/src/autogluon/multimodal/utils.py [hunk 3]
```diff
@@ -2693,3 +2693,40 @@ def from_coco(
...
+            UserWarning,
+        )
+        precision = 32  # Force to use fp32 for training since fp16-based AMP is not available in CPU
+    else:
+        if precision == "bf16" and not torch.cuda.is_bf16_supported():
+            warnings.warn(
+                "bf16 is not supported by the GPU device / cuda version. "
+                "Consider to use GPU devices with version after Amphere or upgrade cuda to be >=11.0. "
```
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/utils.py]: Thanks! Corrected.

**Review by bryanyzhu — APPROVED** [review]
LGTM. We might need to discuss later how to deal with the longer and longer utils.py 😄

**Pull request merged**